src/Dialect/ONNX/IndexExpr.hpp

```diff
@@ -0,0 +1,348 @@
+  bool IsDefined() const;
+  bool IsUndefined() const { return !IsDefined(); }
+  bool IsLiteral() const;
+  bool IsQuestionmark() const;
+  bool IsAffine() const;
+  bool IsSymbol() const;
+  bool IsDim() const;
+  bool IsShapeInferencePass() const;
+  bool HasContext() const;
+  bool HasAffineExpr() const;
+  bool HasValue() const;
+
+  // Getters.
+  int64_t GetLiteral() const;
+  AffineExpr GetAffineExpr();
+  Value GetValue();
+  IndexExprContext &GetContext() const;
+  IndexExprContext *GetContextPtr() const;
+  Location GetLocation() const;
+
+  // Possibly Affine Operations.
+  IndexExpr &Add(IndexExpr &a, IndexExpr &b);
+  IndexExpr &Add(IndexExpr &a, int64_t b);
+  IndexExpr &IncBy(IndexExpr &b);
+  IndexExpr &IncBy(int64_t b);
+  IndexExpr &Sub(IndexExpr &a, IndexExpr &b);
+  IndexExpr &Sub(IndexExpr &a, int64_t b);
+  IndexExpr &Sub(int64_t a, IndexExpr &b);
+  IndexExpr &DecBy(IndexExpr &b);
+  IndexExpr &DecBy(int64_t b);
+  IndexExpr &Mult(IndexExpr &a, IndexExpr &b);
+  IndexExpr &Mult(IndexExpr &a, int64_t b);
+  IndexExpr &MultBy(IndexExpr &b);
+  IndexExpr &MultBy(int64_t b);
+  IndexExpr &FloorDiv(IndexExpr &a, IndexExpr &b);
+  IndexExpr &FloorDivBy(IndexExpr &b);
+  IndexExpr &CeilDiv(IndexExpr &a, IndexExpr &b);
+  IndexExpr &CeilDivBy(IndexExpr &b);
+  IndexExpr &Mod(IndexExpr &a, IndexExpr &b);
+  IndexExpr &ModBy(IndexExpr &a);
+  IndexExpr &Clamp(IndexExpr &val, IndexExpr &min, IndexExpr &max);
+  IndexExpr &Clamp(IndexExpr &val, int64_t min, IndexExpr &max);
+  IndexExpr &Select(IndexExpr &condA, CmpIPredicate comparePred,
+      IndexExpr &condB, IndexExpr &trueVal, IndexExpr &falseVal);
+  IndexExpr &Select(IndexExpr &condA, CmpIPredicate comparePred, int64_t condB,
+      IndexExpr &trueVal, IndexExpr &falseVal);
+  IndexExpr &Select(IndexExpr &condA, CmpIPredicate comparePred, int64_t condB,
+      int64_t trueVal, IndexExpr &falseVal);
+  IndexExpr &AssignIf(IndexExpr &condA, CmpIPredicate comparePred,
+      int64_t condB, IndexExpr &trueVal);
+  IndexExpr &AssignIf(IndexExpr &condA, CmpIPredicate comparePred,
+      int64_t condB, int64_t trueVal);
+  IndexExpr &Min(SmallVectorImpl<IndexExpr> &vals);
+  IndexExpr &Max(SmallVectorImpl<IndexExpr> &vals);
+  void DebugPrint(const std::string &msg);
+
+private:
+  // Higher-level basic initalization calls.
+  IndexExpr &InitAsUndefined();
+  IndexExpr &InitAsQuestionmark(IndexExprContext &context);
+  IndexExpr &InitAsLiteral(IndexExprContext &context, int64_t val);
+  IndexExpr &InitAsSymbol(IndexExprContext &context, Value val);
+  IndexExpr &InitAsDim(IndexExprContext &context, Value val);
+  IndexExpr &InitAsValue(IndexExprContext &context, Value val);
+  IndexExpr &InitAsAffineExpr(IndexExprContext &context, AffineExpr val);
+  // Higher-level initiation calls that extract info
+  IndexExpr &InitAsDimFromMemref(IndexExprContext &context, Value memref,
+      ArrayRef<int64_t> memrefShape, int index);
+  IndexExpr &InitAsSymbolFromArrayAtIndex(
```

Comment:
The term Dim and Index is not very clear to me. 
1. We may want to use a different name for IndexExpr.
2. AtIndex: means for a dimension specified, or the position of the affine map?
3. AsDim: is it for loop index or array dimension?

Reply:
1 & 3) Dim and Symbol are concept from AffineExpr. It is confusing and I don't like it much too. Dim are "runtime input sizes" during shape inferences, and Dim are "loop index" during lowering of code inside the loop iterations. 

I can rename Dim to be (both) "InputDim" and "LoopInductionVar" if you prefer, in the IndexExprContext.create* methods.

2) index of the array. Let me rename the parameter names to make it clearer.